src/nvim/autocmd.c

```diff
@@ -2023,15 +2023,22 @@ char_u *getnextac(int c, void *cookie, int indent, bool do_concat)
   }
 
   if (ac->exec.type == CALLABLE_CB) {
+    dict_T *dict = tv_dict_alloc();
```

Comment:
@gpanders What I'm trying to say is that for every autocmd that matches and is a callback we are allocating memory for the args, providing a unique dictonary to the callback. Perhaps is better to have the memory allocated previously, have only one dictionary and with every pass of the args just rewrite the values that should be on the ditcionary.

If the callback wants to store it, use it later, etc. They should call `deepcopy`.

Reply:
Doesn't Lua create copies anyway?

Reply:
It would require extra glue code, to have a global that we update instead of converting the `typval_T` to a new lua table. Perhaps not worth all the effort? Feel free to dismiss 😄

Reply:
But would it? Whenever you pass anything into Lua, the C API will create copies (AFAIK), so we don't have to worry about using deepcopy or anything, users just need to keep a ref to the table and the Lua-C interface will just figure it out. This typeval stuff is just an additional layer to that.

Reply:
You are right, anything we push into the stack is going to be copied anyway in this case mainly strings if they are not there already.

But I believe in the C side we can avoid allocating a dict every time we are going to pass args and just reuse the same one.

Reply:
I tried this in e9d9caada27913defa0ed167a6407406e29d7563. It took a few other code changes because static dicts aren't really something that has been used before. But I do think this might be worthwhile because autocommand callbacks can potentially fire *a lot* and very quickly, so minimizing memory allocations is a worthwhile goal (there is still required allocations for adding the items to the dict, however). It's going through CI now.